### PR TITLE
[RISCV][CostModel] Correct the cost of some reductions

### PR DESCRIPTION
**Pull request by LiqinWeng**
Reductions include: and/or/max/min

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1543]: Shouldn't this be
```suggestion
      return LT.first * getRISCVInstructionCost(Opcodes, LT.second, CostKind);
```

With that said, I'm also not sure why the existing sequences are adding the legalization cost instead of multiplying it. We multiply it everywhere else in RISCVTargetTransformInfo. Maybe this is something to fix in a follow up patch

Comment (LiqinWeng) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1543]: OK, removed and will create new patch to deal with it

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp
Comment (LiqinWeng) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1473]: done

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp
Comment (LiqinWeng) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1561]: done

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp
Comment (LiqinWeng) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: deal with the type: ElementTy->isIntegerTy(1), should we real need use *???? I don't understand  the case of `ISD == ISD::AND`,  it uses `LT.first - 1`. I'm not sure why it subtracts 1.

Comment (LiqinWeng) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: mycost : LT.first  - 1 + getRISCVInstructionCost + 1(1 is cost of the andi)

Comment (arcbbb) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1543]: `(LT.first - 1)` first appeared in this [commit](https://github.com/llvm/llvm-project/commit/536095a27c17e2f30765111f57d36c0ee9e211c7).
It seems to be used to calculate the additional cost when the LMUL exceeds 8.
For example,
```
define zeroext i1 @vreduce_and_nxv1i1(<vscale x 128 x i1> %v) {
;    vmand.mm    v8, v0, v8 // addtional cost when LMUL exceeds 8.
;    vmnot.m v8, v8
;    vcpop.m a0, v8
;    seqz    a0, a0
  %red = call i1 @llvm.vector.reduce.and.nxv128i1(<vscale x 128 x i1> %v)
  ret i1 %red
}
```

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp
Comment (arcbbb) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1541]: Doesn't the else block already cover this case?

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp
Comment (LiqinWeng) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: thanks~

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp
Comment (LiqinWeng) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: Yes， fixed

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp
Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: Yeah LT.first is the cost of type legalization, i.e. splitting. But it's the number of times the operation will be split, so we need to multiply that by the cost of whatever node. I'll open up a PR to fix the existing cases, but for this PR we might as well do it the correct way i.e. by multiplying. (Although I don't think it will make a difference to the costs, since these mask instructions are always LMUL 1!)

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp
Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: XOR doesn't use a seqz though so it shouldn't use `getCmpSelInstrCost`?

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp
Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: I've gone ahead and done this in df10f1c6f79a77e45e44c62974a186decb0f1f33

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp
Comment (arcbbb) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: @lukel97 You're right. I initially considered replacing getCmpSelInstrCost with 1 to streamline the code. However, I'm fine if you'd prefer to handle the cost separately.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp
Comment (arcbbb) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: Thanks!

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp
Comment (LiqinWeng) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: fixed

**Review by arcbbb — APPROVED** [review]
LGTM.  Please wait for one more approval.

**Review by lukel97 — APPROVED** [review]
LGTM, thanks

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp
Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1487]: For the max/min reductions our lowering just reuses the source vector for the start value, e.g.

```llvm
define signext i8 @vreduce_umax_nxv1i8(<vscale x 1 x i8> %v) {
; CHECK-LABEL: vreduce_umax_nxv1i8:
; CHECK:       # %bb.0:
; CHECK-NEXT:    vsetvli a0, zero, e8, mf8, ta, ma
; CHECK-NEXT:    vredmaxu.vs v8, v8, v8
; CHECK-NEXT:    vmv.x.s a0, v8
; CHECK-NEXT:    ret
  %red = call i8 @llvm.vector.reduce.umax.nxv1i8(<vscale x 1 x i8> %v)
  ret i8 %red
}
```

Comment (preames) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1487]: Oh, so we only removed it for the ones where the start wasn't required.  I hadn't noticed that e.g. ADD wasn't changed.  Make sense - though the comment could definitely be clearer on this point.

Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1487]: Agreed, I had to look back up why this was the case. @LiqinWeng could you update the comment to mention that we reuse the source as the start value for these reductions?

**Pull request merged**
